### PR TITLE
Add support for relative filtering

### PR DESCRIPTION
**Pull request by soshial**
This PR tries to address problem #4735.

### CODE REVIEW

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 1]
```diff
@@ -167,19 +165,56 @@ private boolean filter(Position position) {
...
 
+                Date newfixTime = position.getFixTime();
+                Position duplicate = Context.getDataManager().getPositionByTime(position.getDeviceId(), newfixTime);
+                if (filterDuplicate(position, duplicate) && !skipLimit(position, duplicate)
+                        && !skipAttributes(position)) {
+                    filterType.append("Duplicate ");
+                }
+
+                Position previous = Context.getDataManager().getPrevPosition(position.getDeviceId(), newfixTime);
+                Position next = Context.getDataManager().getNextPosition(position.getDeviceId(), newfixTime);
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:182]: Can you please provide details on why we need "next" position.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: The backend has configured MinPeriod filter set to 2 seconds. Imagine that these  locations were received by the server:
* Location A: `fixTime = 18:18:01`
* Location C: `fixTime = 18:18:05`

Then after 10 seconds, server receives location B with `fixTime = 18:18:04`. As it turns out, it was not delivered immediately because device was offline for 3 seconds, because it was switching between networks.

Location B is far enough from location A, but it's too close to location C time-wise. Hence, we need to take into account _next_ position as well.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: It makes sense in theory, but I'm not sure if it makes sense to actually implement it. You basically make 3 expensive SQL requests for every location report. It will cripple pretty much any server that runs any significant number of devices.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: I agree that with thousands of tracking devices this will have deteriorating impact. But for private use it will not be a problem. No doubt, that we should make this option disabled by default.

What do you say, if I introduce this optimization: let's load _next_ and _prev_ positions only in case any of these filters are turned active:
- filterDistance
- filterMaxSpeed
- filterMinPeriod

So if we only duplicate filter is active, than we have only 1 request instead of 3.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: I'm mostly worried about value of this feature vs cost of maintaining it going forward.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: I thought [you were Okay](https://github.com/traccar/traccar/issues/4735#issuecomment-908474813) with me adding this logic turned OFF by default.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Yeah, the discussion was about loading one last position, if I understood it correctly.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: I can rewrite the code so that it uses only 1 sql query:

`SELECT * FROM tc_positions WHERE deviceId = :deviceId AND NOT fixTime <= :time ORDER BY fixTime DESC  LIMIT 1`

Would it be satisfactory?

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Yes, that makes sense to me. In that case I think the logic would basically stay the same and you just change the way you get "last" position.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Why do you use `last` to calculate speed, when you may take speed value from Position (if available)?

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: It it used for filtering big jumps, like when device suddenly moves to a 0,0 location. Speed reported by device is not useful in those cases.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Can we call `skipLimit` and `skipAttributes` only once in handleFilter()?

```
if (skipLimit(position, previous) || skipAttributes(position)) {
    return true;
}
```

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: We can also make lazy loading of last / previous position. What do you say? Only in the case when at least one of the corresponding filters is activated.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Makes sense to me.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Check this out:
image

line 149: moved skipAttr to the top
line 176: lazy instantiation of _previous_

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:182]: Looks fine, but please update PR so I can review it more thoroughly.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java]: OK, I have pushed a new commit.

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 2]
```diff
@@ -167,21 +168,43 @@ private boolean filter(Position position) {
...
-            filterType.append("Distance ");
-        }
-        if (filterMaxSpeed(position, last)) {
-            filterType.append("MaxSpeed ");
-        }
-        if (filterMinPeriod(position, last)) {
-            filterType.append("MinPeriod ");
+
+        // relative filtering
+        if (filterDuplicate || filterDistance != 0 || filterMaxSpeed != 0 || filterMinPeriod != 0) {
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:176]:
```suggestion
        if (filterDuplicate || filterDistance > 0 || filterMaxSpeed > 0 || filterMinPeriod > 0) {
```

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 3]
```diff
@@ -167,21 +168,43 @@ private boolean filter(Position position) {
...
+            Position previous = null;
+            if (filterRelative) {
+                try {
+                    Date newFixTime = position.getFixTime();
+                    previous = Context.getDataManager().getPrevPosition(position.getDeviceId(), newFixTime);
+                } catch (SQLException e) {
+                    LOGGER.warn("Error filtering position", e);
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:183]: This should probably be some more relevant message, like "Error retrieving previous position" or something along those lines.

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 4]
```diff
@@ -167,21 +168,43 @@ private boolean filter(Position position) {
...
+            if (filterMaxSpeed(position, previous)) {
+                filterType.append("MaxSpeed ");
+            }
+            if (filterMinPeriod(position, previous)) {
+                filterType.append("MinPeriod ");
+            }
         }
 
         if (filterType.length() > 0) {
-
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:184]: undo

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 5]
```diff
@@ -141,23 +146,19 @@ private boolean skipAttributes(Position position) {
     }
 
     private boolean filter(Position position) {
+        if (skipAttributes(position)) {
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:149]:
```suggestion

        if (skipAttributes(position)) {
```

File: src/main/java/org/traccar/handler/FilterHandler.java (hunk 3 above)
Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:183]: We can also fallback to `getLastPosition`.

File: src/main/java/org/traccar/config/Keys.java [hunk 6]
```diff
@@ -887,15 +890,27 @@ private Keys() {
             Collections.singletonList(KeyType.GLOBAL));
 
     /**
-     * Time limit for the filtering in seconds. If the time difference between last position and a new one is more than
-     * this limit, the new position will not be filtered out.
+     * If FALSE, then filter compares duplicates, distance or period only with the last received location.
+     * Server expects all locations to come sequentially.
+     *
+     * If TRUE, then filter compares duplicates, distance or period with Position's fixTime strictly before and after.
+     * Server expects locations to come at random order, since tracking device might go offline.
```
Comment (tananaev) [inline src/main/java/org/traccar/config/Keys.java:897]: For consistency
```suggestion
     * If false, then filter compares duplicates, distance or period only with the last received location.
     * Server expects all locations to come sequentially.
     *
     * If true, then filter compares duplicates, distance or period with Position's fixTime strictly before and after.
     * Server expects locations to come at random order, since tracking device might go offline.
```

File: src/main/java/org/traccar/database/DataManager.java [hunk 7]
```diff
@@ -328,6 +328,42 @@ public void updateDeviceStatus(Device device) throws SQLException {
...
+                .setDate("time", date)
+                .executeQuery(Position.class);
+        if (positions.isEmpty()) {
+            return null;
+        } else {
+            return positions.iterator().next();
+        }
+    }
+
+    public Position getNextPosition(long deviceId, Date date) throws SQLException {
```
Comment (tananaev) [inline src/main/java/org/traccar/database/DataManager.java:355]: We only need one of these now, right? Also, please use full words, like `previous` instead of `prev`.

File: src/main/java/org/traccar/config/Keys.java (hunk 6 above)
Comment (tananaev) [reply src/main/java/org/traccar/config/Keys.java:897]: Can you also add a quick note that setting to true can cause potential performance issues.

File: setup/default.xml [hunk 8]
```diff
@@ -22,6 +22,7 @@
 
     <entry key='filter.enable'>true</entry>
     <entry key='filter.future'>86400</entry>
+    <entry key='filter.relative'>false</entry>
```
Comment (tananaev) [inline setup/default.xml:25]: Don't need to add this if it's already the default.

**Review by tananaev — COMMENTED** [review]
Looks good. Add a few minor comments, mostly about formatting.

File: src/main/java/org/traccar/database/DataManager.java (hunk 7 above)
Comment (soshial) [reply src/main/java/org/traccar/database/DataManager.java:355]: Can I rename it everywhere to `preceeding`? It's much clearer title

Comment (tananaev) [reply src/main/java/org/traccar/database/DataManager.java:355]: Sounds good

File: setup/default.xml [hunk 9]
```diff
@@ -48,6 +48,14 @@
         SELECT * FROM tc_positions WHERE deviceId = :deviceId AND fixTime BETWEEN :from AND :to ORDER BY fixTime
     </entry>
 
+    <entry key='database.selectPositionByTime'>
```
Comment (tananaev) [inline setup/default.xml:51]: This one should be removed as well, right?

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 10]
```diff
@@ -142,22 +147,19 @@ private boolean skipAttributes(Position position) {
...
-        Position last = null;
-        if (Context.getIdentityManager() != null) {
-            last = Context.getIdentityManager().getLastPosition(position.getDeviceId());
+        if (skipAttributes(position)) {
+            return false;
         }
 
+        StringBuilder filterType = new StringBuilder();
+
+        // general filtering
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:156]: Can you also please remove this comment.

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 11]
```diff
@@ -167,17 +169,40 @@ private boolean filter(Position position) {
...
-        if (filterDistance(position, last) && !skipLimit(position, last) && !skipAttributes(position)) {
-            filterType.append("Distance ");
-        }
-        if (filterMaxSpeed(position, last)) {
-            filterType.append("MaxSpeed ");
-        }
-        if (filterMinPeriod(position, last)) {
-            filterType.append("MinPeriod ");
+
+        // relative filtering
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:176]: And this.

File: setup/default.xml (hunk 9 above)
Comment (tananaev) [reply setup/default.xml:51]: This is not resolved yet, as far as I can tell.

**Comment by soshial** [comment]
It is. I force pushed the commit.

File: setup/default.xml [hunk 12]
```diff
@@ -48,6 +48,10 @@
         SELECT * FROM tc_positions WHERE deviceId = :deviceId AND fixTime BETWEEN :from AND :to ORDER BY fixTime
     </entry>
 
+    <entry key='database.selectPrecedingPosition'>
+        SELECT * FROM tc_positions WHERE deviceId = :deviceId AND fixTime lte :time ORDER BY fixTime DESC LIMIT 1
```
Comment (tananaev) [inline setup/default.xml:52]: One last thing. I noticed that you are using `lte`. Is it a standard SQL syntax that will work in every SQL server? Can you please provide a reference because I couldn't find any information about it.

Comment (soshial) [reply setup/default.xml:52]: Honestly, I thought that you were using some kind of library to parse SQL stored in XML. Did you figure out correctly, that you wrote it yourself. I couldn't use `<` sign since it's forbidden inside XML tag.

We may use this:
`SELECT * FROM tc_positions WHERE deviceId = :deviceId AND fixTime &lt;= :time ORDER BY fixTime DESC LIMIT 1`
or
`SELECT * FROM tc_positions WHERE deviceId = :deviceId AND NOT (fixTime > :time) ORDER BY fixTime DESC LIMIT 1`

What do you think?

File: src/main/java/org/traccar/database/DataManager.java [hunk 13]
```diff
@@ -328,6 +328,18 @@ public void updateDeviceStatus(Device device) throws SQLException {
...
+    public Position getPrecedingPosition(long deviceId, Date date) throws SQLException {
+        Collection<Position> positions = QueryBuilder.create(dataSource, getQuery("database.selectPrecedingPosition"))
+                .setLong("deviceId", deviceId)
+                .setDate("time", date)
+                .executeQuery(Position.class);
+        if (positions.isEmpty()) {
+            return null;
+        } else {
+            return positions.iterator().next();
+        }
```
Comment (tananaev) [inline src/main/java/org/traccar/database/DataManager.java:340]: Sorry, just noticed another thing. Can we use `executeQuerySingle` here instead?

**Review by tananaev — COMMENTED** [review]
Also, just to confirm. Have you tested the latest to make sure everything works?

**Comment by soshial** [comment]
I have tested the code and I found a serious problem (as it seems to me), that have existed in traccar even before my pull request. As it turns out, traccar doesn't save fractional TIMESTAMP values, it saves only seconds (not milliseconds). Which means that duplicate comparison didn't work for protocols that send milliseconds:

`position.getFixTime().getTime() == last.getFixTime().getTime()` returns false for seemingly duplicate position (that I send repeatedly). When comparing to last received position (as was implemented before) this problem did not appear. But ifwe take values from DB, it becomes a problem.

Compare:
* the new received position had value `1626692627800`
* and the position from DB has value `1626692627000`

**Comment by tananaev** [comment]
What database are you using?

**Comment by soshial** [comment]
MariaDB.

**Comment by tananaev** [comment]
Is there a universal solution?

**Comment by soshial** [comment]
What do you mean?

**Comment by tananaev** [comment]
I mean any ideas how to fix it via liquibase, so it's supported for add databases?

**Comment by soshial** [comment]
Looks like this should do the thing:

    <modifyDataType tableName="tc_positions" columnName="fixtime" newDataType="TIMESTAMP(6)" />
    <modifyDataType tableName="tc_positions" columnName="servertime" newDataType="TIMESTAMP(6)" />
    <modifyDataType tableName="tc_positions" columnName="devicetime" newDataType="TIMESTAMP(6)" />
    <modifyDataType tableName="tc_events" columnName="servertime" newDataType="TIMESTAMP(6)" />
    <modifyDataType tableName="tc_devices" columnName="lastupdate" newDataType="TIMESTAMP(6)" />

How do check that it works?

**Comment by tananaev** [comment]
You have to check if it works with different database types. At least H2, MySQL, MS SQL and PostgreSQL.

**Comment by soshial** [comment]
Let's switch to another PR? (since it's a different bug fix)

**Comment by tananaev** [comment]
Yep, makes sense to do a separate PR.

**Comment by soshial** [comment]
Besides this separate time issue, I would say everything works for me.

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 14]
```diff
@@ -54,22 +58,23 @@ public FilterHandler(Config config) {
...
+                || position.getLatitude() < -90 || position.getLongitude() < -180);
     }
 
     private boolean filterZero(Position position) {
         return filterZero && position.getLatitude() == 0.0 && position.getLongitude() == 0.0;
     }
 
     private boolean filterDuplicate(Position position, Position last) {
-        if (filterDuplicate && last != null && position.getFixTime().equals(last.getFixTime())) {
+        if (filterDuplicate && last != null && position.getFixTime().getTime() == last.getFixTime().getTime()) {
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:77]: Why is this needed?

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: If we're going to store milli/micro/nanoseconds in the TIMESTAMP field in the DB, then it means we would in fact compare float numbers. So maybe to filter out duplicates we should compare `getTime()` with a specified precision.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: Sorry, where did you get floating numbers and nanoseconds? As far as I can tell, `equals` does exactly the same thing as you do manually:

https://docs.oracle.com/javase/8/docs/api/java/util/Date.html#equals-java.lang.Object-

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: Yes, currently it does the same thing. But, if we make this change https://github.com/traccar/traccar/pull/4746, then database would store time with milliseconds/nanosecond precision (as float). Hence, comparing new position (converted from text) and preceding (from DB) might not be equal in theory. That is why I suggest to do comparison with a certain precision. I can do additional commit for that. Do you agree?

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: I thought you planned to fix database accuracy?

**Comment by tsmgeek** [comment]
Note with moving to fractional seconds you may need to account for `BETWEEN` queries (from UI etc) that may be off by 0.999999 seconds.
Ie between 00:01-00:02 when actually you may have meant `00:01.000000-00:02.999999`.

In some apps at work we just use https://docs.oracle.com/javase/8/docs/api/java/time/LocalDateTime.html#MAX in code to always push max microseconds to account for this.

File: src/main/java/org/traccar/handler/FilterHandler.java (hunk 14 above)
Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: Let's leave it as it is now. And deal with it in the separate PR.

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:77]: Agree but this change should be reverted because it doesn't really do anything.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java]: Done.

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 15]
```diff
@@ -142,22 +147,18 @@ private boolean skipAttributes(Position position) {
 
     private boolean filter(Position position) {
 
-        StringBuilder filterType = new StringBuilder();
-
-        Position last = null;
-        if (Context.getIdentityManager() != null) {
-            last = Context.getIdentityManager().getLastPosition(position.getDeviceId());
+        if (skipAttributes(position)) {
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:150]: Looks like this actually changes the filtering logic. We only want to apply `skipLimit` and `skipAttributes` for certain types of filters.

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java:150]: Okay. Could you please elaborate? Why should those 2 _skips_ work only for some filters?

Comment (tananaev) [reply src/main/java/org/traccar/handler/FilterHandler.java:150]: There was a forum discussion about it if you want more context. Basically there are two categories of filtering. The ones that filter definitely invalid data (e.g. zero coordinates) and the ones that filter excessive data (e.g. duplicates). We always want to filter former, but want to keep latter if it has some important information (e.g. alarms).

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java]: Done. I reverted to the original logic and also preserved almost the same order of filter checks (as it is in current master branch).

File: src/main/java/org/traccar/handler/FilterHandler.java [hunk 16]
```diff
@@ -167,17 +165,37 @@ private boolean filter(Position position) {
...
-        }
-        if (filterMaxSpeed(position, last)) {
-            filterType.append("MaxSpeed ");
-        }
-        if (filterMinPeriod(position, last)) {
-            filterType.append("MinPeriod ");
+
+        // filter out excessive data
+        long deviceId = position.getDeviceId();
+        if (filterDuplicate || filterDistance > 0 || filterMaxSpeed > 0 || filterMinPeriod > 0) {
```
Comment (tananaev) [inline src/main/java/org/traccar/handler/FilterHandler.java:171]: What about `filterStatic`?

Comment (soshial) [reply src/main/java/org/traccar/handler/FilterHandler.java]: Fixed.

**Comment by tananaev** [comment]
Merged, thank you.

**Comment by soshial** [comment]
Could you please add the new `filter.relative` key to [the documentation](https://www.traccar.org/configuration-file/)?

**Comment by tananaev** [comment]
The plan is to replace that documentation with something that's generated automatically.
